### PR TITLE
Add Network Acl Attachment Resource

### PR DESCRIPTION
**Pull request by dianmuali**
MacBook-Pro-2:terraform-provider-alicloud jince$ TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudNetworkAcl -timeout=300m
=== RUN   TestAccAlicloudNetworkAcl_importBasic
--- SKIP: TestAccAlicloudNetworkAcl_importBasic (0.00s)
    provider_test.go:90: Skipping unsupported region cn-beijing. Supported regions: [cn-hongkong ap-southeast-5 ap-south-1].
=== RUN   TestAccAlicloudNetworkAclAttachment_basic
--- SKIP: TestAccAlicloudNetworkAclAttachment_basic (0.00s)
    provider_test.go:90: Skipping unsupported region cn-beijing. Supported regions: [cn-hongkong ap-southeast-5 ap-south-1].
=== RUN   TestAccAlicloudNetworkAcl_basic
--- SKIP: TestAccAlicloudNetworkAcl_basic (0.00s)
    provider_test.go:90: Skipping unsupported region cn-beijing. Supported regions: [cn-hongkong ap-southeast-5 ap-south-1].
=== RUN   TestAccAlicloudNetworkAcl_multi
--- SKIP: TestAccAlicloudNetworkAcl_multi (0.00s)
    provider_test.go:90: Skipping unsupported region cn-beijing. Supported regions: [cn-hongkong ap-southeast-5 ap-south-1].
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	0.052s

### CODE REVIEW

File: alicloud/service_alicloud_network_acl.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	response, _ := raw.(*vpc.DescribeNetworkAclsResponse)
+	addDebug(request.GetActionName(), raw)
+	if len(response.NetworkAcls.NetworkAcl) <= 0 || response.NetworkAcls.NetworkAcl[0].NetworkAclId != id {
+		return networkAcl, WrapErrorf(Error(GetNotFoundMessage("Network Acl", id)), NotFoundMsg, ProviderERROR)
+	}
+	return response.NetworkAcls.NetworkAcl[0], nil
+}
+
+func (s *NetworkAclService) DescribeNetworkAclAttachment(id string, resource []vpc.Resource) (err error) {
```
Comment (ddaddaddda) [inline alicloud/service_alicloud_network_acl.go]: 1 建议改一下attachment 的 id，将acl 的id 与绑定的资源类似id的属性组合一下，describe方法只使用id一个入参，使acl与绑定的资源相关联，方便校验绑定是否成功；
2 返回参数需要两个，vpc.NetworkAcl 与 error 即可，方便read方法与测试checkExist调用；
3 DescribeNetworkAcl 方法可在本方法中复用。

可参照  alicloud_key_pair_attachment

File: alicloud/provider.go [hunk 2]
```diff
@@ -306,6 +306,8 @@ func Provider() terraform.ResourceProvider {
 			"alicloud_actiontrail":                         resourceAlicloudActiontrail(),
 			"alicloud_cas_certificate":                     resourceAlicloudCasCertificate(),
 			"alicloud_ddoscoo_instance":                    resourceAlicloudDdoscooInstance(),
+			"alicloud_network_acl":                         resourceAliyunNetworkAcl(),
```
Comment (ddaddaddda) [inline alicloud/provider.go]: 309   pr  1151 里面应该加过 "alicloud_network_acl" 了吧

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	request.NetworkAclId = d.Id()
+	request.Resource = &resources
+	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+		raw, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+			return vpcClient.UnassociateNetworkAcl(request)
+		})
+		//Waiting for unassociate the network acl
+		if err != nil {
+			if IsExceptedError(err, TaskConflict) {
+				return resource.RetryableError(fmt.Errorf("Unassociate Network Acl timeout and got an error:%#v.", err))
```
Comment (ddaddaddda) [inline alicloud/resource_alicloud_network_acl_attachment.go]: 146  resource.RetryableError(err)

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			}
+		}
+		addDebug(request.GetActionName, raw)
+		return nil
+	})
+	err = networkAclService.WaitForNetworkAclAttachment(d.Id(), vpcResource, Deleted, DefaultTimeout)
```
Comment (ddaddaddda) [inline alicloud/resource_alicloud_network_acl_attachment.go]: 152 这个应该删除的

File: alicloud/resource_alicloud_network_acl_attachment_test.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+		}
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No Network Acl ID is set")
+		}
+		client := testAccProvider.Meta().(*connectivity.AliyunClient)
+		networkAclService := NetworkAclService{client}
+		object, err := networkAclService.DescribeNetworkAcl(rs.Primary.ID)
+		err = networkAclService.DescribeNetworkAclAttachment(rs.Primary.ID, object.Resources.Resource)
+		if err != nil {
+			return fmt.Errorf("Describe Route Table attachment error %#v", err)
```
Comment (ddaddaddda) [inline alicloud/resource_alicloud_network_acl_attachment_test.go]: 151 return WrapError(err)就行

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		if NotFoundError(err) {
+			return nil
+		}
+		return WrapError(err)
```
Comment (ddaddaddda) [inline alicloud/resource_alicloud_network_acl_attachment.go]: 157 用WrapErrorf

File: alicloud/service_alicloud_network_acl.go [hunk 7]
```diff
@@ -33,6 +33,33 @@ func (s *NetworkAclService) DescribeNetworkAcl(id string) (networkAcl vpc.Networ
...
+
+	invoker := NewInvoker()
+	return invoker.Run(func() error {
+		object, err := s.DescribeNetworkAcl(id)
+		if err != nil {
+			return err
+		}
+		if len(object.Resources.Resource) < 1 {
+			return WrapErrorf(Error(GetNotFoundMessage("Network Acl", id)), NotFoundMsg, ProviderERROR)
```
Comment (ddaddaddda) [inline alicloud/service_alicloud_network_acl.go]: 45  Network Acl -> NetworkAclAttachment

File: alicloud/service_alicloud_network_acl.go [hunk 8]
```diff
@@ -33,6 +33,33 @@ func (s *NetworkAclService) DescribeNetworkAcl(id string) (networkAcl vpc.Networ
...
+		success := true
+		for _, source := range object.Resources.Resource {
+			success = false
+			for _, res := range resource {
+				if source.ResourceId == res.ResourceId {
+					success = true
+				}
+			}
+			if success == false {
+				return WrapErrorf(Error(GetNotFoundMessage("Network Acl", id)), NotFoundMsg, ProviderERROR)
```
Comment (ddaddaddda) [inline alicloud/service_alicloud_network_acl.go]: 56 同上

File: alicloud/service_alicloud_network_acl.go [hunk 9]
```diff
@@ -55,3 +82,24 @@ func (s *NetworkAclService) WaitForNetworkAcl(networkAclId string, status Status
...
+	for {
+		err := s.DescribeNetworkAclAttachment(id, resource)
+		if err != nil {
+			if NotFoundError(err) {
+				return WrapErrorf(Error(GetNotFoundMessage("Network Acl", id)), NotFoundMsg, ProviderERROR)
+			} else {
+				return WrapError(err)
+			}
+		}
+		if status == Available {
```
Comment (ddaddaddda) [inline alicloud/service_alicloud_network_acl.go]: 97  用acl的status判断

File: alicloud/service_alicloud_network_acl.go [hunk 10]
```diff
@@ -55,3 +82,24 @@ func (s *NetworkAclService) WaitForNetworkAcl(networkAclId string, status Status
...
 }
+
+func (s *NetworkAclService) WaitForNetworkAclAttachment(id string, resource []vpc.Resource, status Status, timeout int) error {
+	deadline := time.Now().Add(time.Duration(timeout) * time.Second)
+	for {
+		err := s.DescribeNetworkAclAttachment(id, resource)
+		if err != nil {
+			if NotFoundError(err) {
+				return WrapErrorf(Error(GetNotFoundMessage("Network Acl", id)), NotFoundMsg, ProviderERROR)
+			} else {
```
Comment (ddaddaddda) [inline alicloud/service_alicloud_network_acl.go]: 92 if status == Delete{ return nil}

File: alicloud/resource_alicloud_network_acl_attachment_test.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+			},
+		},
+	})
+}
+
+func testAccCheckNetworkAclAttachmentExists(n string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
```
Comment (ddaddaddda) [inline alicloud/resource_alicloud_network_acl_attachment_test.go]: 141  return WrapError(Error("Not found: %s", n))

File: alicloud/resource_alicloud_network_acl_attachment_test.go [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+		}
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No Network Acl ID is set")
```
Comment (ddaddaddda) [inline alicloud/resource_alicloud_network_acl_attachment_test.go]: 144  return WrapError(Error("No Network Acl ID is set"))

File: alicloud/resource_alicloud_network_acl_attachment_test.go [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+			vpcSource := vpc.Resource{
+				ResourceId:   e.ResourceId,
+				ResourceType: e.ResourceType,
+			}
+			vpcResource = append(vpcResource, vpcSource)
+		}
+		err = networkAclService.WaitForNetworkAclAttachment(rs.Primary.ID, vpcResource, Deleted, DefaultTimeout)
+		if err != nil {
+			if NotFoundError(err) {
+				return nil
```
Comment (ddaddaddda) [inline alicloud/resource_alicloud_network_acl_attachment_test.go]: 178 continue

File: website/docs/r/network_acl_attachment.html.markdown [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+	vpc_id = "${alicloud_vpc.default.id}"
+	cidr_block = "172.16.0.0/21"
+	availability_zone = "${data.alicloud_zones.default.zones.0.id}"
+	name = "${var.name}"
+}
+
+resource "alicloud_network_acl_attachment" "default" {
+	network_acl_id = "${alicloud_network_acl.default.id}"
+    resources = [
+        {
```
Comment (ddaddaddda) [inline website/docs/r/network_acl_attachment.html.markdown]: 51  调一下缩进

Comment (ddaddaddda) [inline website/docs/r/network_acl_attachment.html.markdown]: 51  调一下缩进

Comment (ddaddaddda) [inline website/docs/r/network_acl_attachment.html.markdown]: 51  调一下缩进

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+				ForceNew: true,
+			},
+			"resources": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"resource_id": {
+							Type:     schema.TypeString,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: Please use `TypeSet`

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+	vpcResource := []vpc.Resource{}
+	for _, e := range d.Get("resources").([]interface{}) {
+		res := e.(map[string]interface{})
+		vpcSource := vpc.Resource{
+			ResourceId:   res["resource_id"].(string),
+			ResourceType: res["resource_type"].(string),
+		}
+		vpcResource = append(vpcResource, vpcSource)
+	}
+	err = networkAclService.DescribeNetworkAclAttachment(object.NetworkAclId, vpcResource)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: I think the line 93 should be moved into the method `DescribeNetworkAclAttachment ` and this method should return a ACL Object.

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+			if IsExceptedError(err, TaskConflict) {
+				return resource.RetryableError(err)
+			}
+		}
+		addDebug(request.GetActionName, raw)
+		return nil
+	})
+	err = networkAclService.WaitForNetworkAclAttachment(d.Id(), vpcResource, Deleted, DefaultTimeout)
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: If the attachment has been deleted, will the method `UnassociateNetworkAcl` return an error, like "Instance not found"? If so, please check it in the lien 153

File: alicloud/resource_alicloud_network_acl_attachment_test.go [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+	request := vpc.CreateDescribeNetworkAclsRequest()
+	request.RegionId = client.RegionId
+	request.PageSize = requests.NewInteger(PageSizeLarge)
+	request.PageNumber = requests.NewInteger(1)
+	for {
+		raw, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+			return vpcClient.DescribeNetworkAcls(request)
+		})
+		if err != nil {
+			return WrapError(err)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment_test.go]: There don't suggest to return a error in the sweeper test and just print it.

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-alicloud/alicloud/connectivity"
+)
+
+func resourceAliyunNetworkAclAttachment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAliyunNetworkAclAttachmentCreate,
+		Read:   resourceAliyunNetworkAclAttachmentRead,
+		Delete: resourceAliyunNetworkAclAttachmentDelete,
+		Importer: &schema.ResourceImporter{
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: If the resource supports `Importer`, please add its testcase.

File: alicloud/resource_alicloud_network_acl_attachment_test.go [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+		vpcResource := []vpc.Resource{}
+		for _, e := range object.Resources.Resource {
+
+			vpcSource := vpc.Resource{
+				ResourceId:   e.ResourceId,
+				ResourceType: e.ResourceType,
+			}
+			vpcResource = append(vpcResource, vpcSource)
+		}
+		err = networkAclService.WaitForNetworkAclAttachment(rs.Primary.ID, vpcResource, Deleted, DefaultTimeout)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment_test.go]: I think there is no need to use  `WaitForNetworkAclAttachment `

File: website/docs/r/network_acl_attachment.html.markdown [hunk 21]
```diff
@@ -0,0 +1,77 @@
+---
+layout: "alicloud"
+page_title: "Alicloud: alicloud_network_acl_attachment"
+sidebar_current: "docs-alicloud-resource-network-acl-attachment"
+description: |-
+  Provides a Alicloud Network Acl Attachment resource.
+---
+
+# alicloud\_network_acl
```
Comment (xiaozhu36) [inline website/docs/r/network_acl_attachment.html.markdown]: please use the correct resource name

File: website/docs/r/network_acl_attachment.html.markdown [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+Provides a network acl attachment resource to associate network acls to vswitches.
+
+-> **NOTE:** Available in 1.44.0. Currently, the resource are only available in Hongkong(cn-hongkong), India(ap-south-1), and Indonesia(ap-southeast-1) regions.
```
Comment (xiaozhu36) [inline website/docs/r/network_acl_attachment.html.markdown]: `Available in 1.44.0+`

File: website/docs/r/network_acl_attachment.html.markdown [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+* `resource_id` - (Required) The resource id that the network acl will associate with.
+* `resource_type` - (Required) The resource id that the network acl will associate with. Only support `VSwitch` now.
+
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The ID of the network acl instance id.
+
```
Comment (xiaozhu36) [inline website/docs/r/network_acl_attachment.html.markdown]: If it supports `Importer`, please show it in the docs.

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 24]
```diff
@@ -0,0 +1,228 @@
...
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"resource_type": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+				MaxItems: 20,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: Please remove the MaxItems if it is useless.

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+func resourceAliyunNetworkAclAttachmentRead(d *schema.ResourceData, meta interface{}) error {
+
+	client := meta.(*connectivity.AliyunClient)
+	networkAclService := NetworkAclService{client}
+	parts, err := ParseResourceId(d.Id(), 2)
+	if err != nil {
+		return WrapError(err)
+	}
+	networkAclId := parts[0]
+	object, err := networkAclService.DescribeNetworkAcl(networkAclId)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: I think the line 63 is needless.

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+				}
+				resources = append(resources, vpc.UnassociateNetworkAclResource{
+					ResourceId:   resourceId,
+					ResourceType: resourceType,
+				})
+				vpcSource := vpc.Resource{
+					ResourceId:   resourceId,
+					ResourceType: resourceType,
+				}
+				vpcResource = append(vpcResource, vpcSource)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: I think the line 120 and 124 can be merged

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+		vpcSource := vpc.Resource{
+			ResourceId:   e.ResourceId,
+			ResourceType: e.ResourceType,
+		}
+		vpcResource = append(vpcResource, vpcSource)
+		source := vpc.UnassociateNetworkAclResource{
+			ResourceId:   e.ResourceId,
+			ResourceType: e.ResourceType,
+		}
+		resources = append(resources, source)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: I think line 203 and 207 can be merged.

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+	}
+	request.Resource = &resources
+	err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+		raw, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+			return vpcClient.UnassociateNetworkAcl(request)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: If the api supports clientToken, please set it.

File: alicloud/service_alicloud_network_acl.go [hunk 29]
```diff
@@ -33,6 +33,33 @@ func (s *NetworkAclService) DescribeNetworkAcl(id string) (networkAcl vpc.Networ
 	return response.NetworkAcls.NetworkAcl[0], nil
 }
 
+func (s *NetworkAclService) DescribeNetworkAclAttachment(id string, resource []vpc.Resource) (err error) {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_network_acl.go]: I think using a map to store the resources will be better.

File: website/docs/r/network_acl_attachment.html.markdown [hunk 30]
```diff
@@ -0,0 +1,77 @@
...
+* `resource_id` - (Required) The resource id that the network acl will associate with.
+* `resource_type` - (Required) The resource id that the network acl will associate with. Only support `VSwitch` now.
+
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The ID of the network acl instance id.
```
Comment (xiaozhu36) [inline website/docs/r/network_acl_attachment.html.markdown]: * `id`` - The ID of the network acl attachment. It is formatted as < network_acl_id >:<a unique id>`.

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 31]
```diff
@@ -0,0 +1,223 @@
...
+					ResourceId:   resourceId,
+					ResourceType: resourceType,
+				})
+			}
+			request.Resource = &resources
+			raw, err := client.WithVpcClient(func(vpcClient *vpc.Client) (interface{}, error) {
+				return vpcClient.AssociateNetworkAcl(request)
+			})
+			if err != nil {
+				return WrapErrorf(err, DefaultErrorMsg, "alicloud_network_acl_attachment", request.GetActionName(), AlibabaCloudSdkGoERROR)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: alicloud_network_acl_attachment -> d.Id()

File: alicloud/resource_alicloud_network_acl_attachment.go [hunk 32]
```diff
@@ -0,0 +1,223 @@
...
+		if err != nil {
+			if IsExceptedError(err, TaskConflict) {
+				return resource.RetryableError(err)
+			}
+		}
+		addDebug(request.GetActionName, raw)
+		return nil
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, networkAclId, request.GetActionName(), AlibabaCloudSdkGoERROR)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_network_acl_attachment.go]: networkAclId -> d.Id()

File: alicloud/service_alicloud_network_acl.go [hunk 33]
```diff
@@ -33,6 +33,33 @@ func (s *NetworkAclService) DescribeNetworkAcl(id string) (networkAcl vpc.Networ
...
+
+	invoker := NewInvoker()
+	return invoker.Run(func() error {
+		object, err := s.DescribeNetworkAcl(id)
+		if err != nil {
+			return err
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_network_acl.go]: return err -> return WrapError(err)

File: website/docs/r/network_acl_attachment.html.markdown [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+  ]
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `network_acl_id` - (Required, ForceNew) The id of the network acl, the field can't be changed.
+* `resources` - (Required) The resources of the network acl,  the field can't be changed.
```
Comment (xiaozhu36) [inline website/docs/r/network_acl_attachment.html.markdown]: "The resources of the network acl,  the field can't be changed." -> List of the resources associated with the network acl. The details see `Block Resources`.

File: website/docs/r/network_acl_attachment.html.markdown [hunk 35]
```diff
@@ -0,0 +1,77 @@
...
+
+### Block Resources
+The resources mapping supports the following:
+
+* `resource_id` - (Required) The resource id that the network acl will associate with.
+* `resource_type` - (Required) The resource id that the network acl will associate with. Only support `VSwitch` now.
+
+
+## Attributes Reference
```
Comment (xiaozhu36) [inline website/docs/r/network_acl_attachment.html.markdown]: Please keep only one blank line between line 68 to 71.

**Pull request merged**